prov/shm/src/smr_ep.c

```diff
@@ -635,8 +635,12 @@ static int smr_format_sar(struct smr_ep *ep, struct smr_cmd *cmd,
 }
 
 int smr_select_proto(enum fi_hmem_iface iface, bool use_ipc, bool cma_avail,
-                     uint32_t op, uint64_t total_len, uint64_t op_flags)
+                     bool gdrcopy_avail, uint32_t op, uint64_t total_len,
+                     uint64_t op_flags)
 {
+	if (gdrcopy_avail && total_len <= smr_env.max_gdrcopy_size)
```

Comment:
Not sure what the end vision is for this but can we just initialize max_grdcopy_size to either 3072 or 0 whether gdrcopy is available or not? Is gdrcopy_avail a global setting or per transfer? 
Eventually I think we should think about reordering this proto selection code to optimize for the fast path

Reply:
My current solution is to have a static threshold 3072, plus a flag `OFI_HMEM_DATA_GDRCOPY_HANDLE` that's passed from EFA to `fi_mr_regattr`.

In e.g. `smr_generic_sendmsg` we will inspect `smr_desc = (struct ofi_mr *) *desc;` and set `gdrcopy_avail` if the flag is set.

So overall the idea is to separate 1) whether gdrcopy handle is available(based on flag) from 2) when gdrcopy is preferred over cudaMemcpy(based on smr env).

> Eventually I think we should think about reordering this proto selection code to optimize for the fast path

Yes. I agree. With this change I do see that we add an additional check `gdrcopy_avail` to all paths. Let me think about this for a min.

Reply:
@aingerson It took me more than 1min. Sadly it eludes me how to do it cleanly... I messed around the code and ended up with plugging in `gdrcopy_avail` left and right.

That said, I tried to understand the impact of this change but so far it appears benign 🫤

|  pt2pt/osu_latency  |                   |                   |
|---------------------|-------------------|-------------------|
|                Size |           Before  |            After  |
|                   0 |             0.95  |             0.93  |
|                   1 |             0.97  |             0.93  |
|                   2 |             0.96  |             0.91  |
|                   4 |             0.96  |             0.92  |
|                   8 |             0.95  |             0.92  |
|                  16 |             0.95  |             0.92  |
|                  32 |             0.96  |             0.94  |
|                  64 |             0.97  |             0.93  |
|                 128 |             1.02  |             0.97  |
|                 256 |             1.16  |             1.13  |
|                 512 |             1.21  |             1.17  |
|                1024 |             1.30  |             1.27  |
|                2048 |             1.47  |             1.44  |
|                2560 |             1.57  |             1.53  |
|                3072 |             1.65  |             1.62  |
|                3584 |             1.70  |             1.69  |
|                4096 |             1.85  |             1.82  |
|                8192 |             2.73  |             2.68  |
|               16384 |             4.02  |             4.00  |
|               32768 |             6.05  |             6.01  |
|               65536 |             9.94  |             9.96  |
|              131072 |            18.48  |            18.47  |
|              262144 |            35.04  |            34.80  |
|              524288 |            80.30  |            79.97  |
|             1048576 |           171.43  |           172.10  |
|             2097152 |           338.99  |           339.70  |

Reply:
No worries. Thanks for taking a look! We can just put it on a future todo to see if we can optimize this.

Reply:
Good idea. Added a TODO comment.